src/anyio/streams/memory.py

```diff
@@ -230,6 +236,12 @@ async def send(self, item: T_contra) -> None:
             self._state.waiting_senders[send_event] = item
             try:
                 await send_event.wait()
+            except get_cancelled_exc_class():
+                # Ignore the immediate cancellation if we already sent the item, so as
+                # to not indicate failure despite success
+                if send_event in self._state.waiting_senders:
+                    del self._state.waiting_senders[send_event]
+                    raise
```

Comment:
This would be fine when the cancellation originates from a cancel scope, but not when a task was natively cancelled (`Task.cancel()`, because the originator of that cancellation expects the cancellation to be acted on and not ignored). Doing this will lead to hangs.